main/src/library/Fixpoint/Solver.flix

```diff
@@ -40,12 +40,19 @@ namespace Fixpoint {
     ///
     pub def solveWithStratification(d: Datalog[v], stf: Map[PredSym, Int32]): Datalog[v] with Order[v], Eq[v], ToString[v] =
         let _ = notifyPreSolve(d, stf);
-        let model = (compile(d, stf) |>
-         simplifyStmt |>
-         queryStmt |>
-         lowerStmt |>
-         interpret |>
-         toAst) as & Pure;
+        let compiler = cs ->
+            compile(cs, stf) |>
+            simplifyStmt |>
+            queryStmt |>
+            lowerStmt;
+        let model = match d {
+            case Datalog(_) => compiler(d) |> interpret |> toModel
+            case Model(_) => d
+            case Join(Model(m), cs) =>
+                let db = Map.map(Map.toMut, m) |> Map.toMut;
+                compiler(cs) |> interpretWithDatabase(db) |> toModel
+            case _ => bug!("Datalog normalization bug")
```

Comment:
Can this case happen?

Reply:
It shouldn't. The logic in `union` is supposed to guarantee the composition is either:
 - A `Datalog` value (if both inputs are `Datalog` values).
 - A `Model` (if both inputs are models)
 - A `Join(Model, Datalog)` (if a `Model` and `Datalog` value is being composed, regardless of the order of composition).